File: home/ubuntu/remitflow/54remitflow-unified-platform/backend/resilient-clients/tigerbeetle/tigerbeetle-resilient-client.ts
/**
 * TigerBeetle Resilient Client
 * 
 * Implements retry logic with exponential backoff and circuit breaker pattern
 * for reliable TigerBeetle operations in production environments.
 * 
 * Features:
 * - Exponential backoff with jitter
 * - Circuit breaker pattern (Closed, Open, Half-Open states)
 * - Configurable timeout policies
 * - Comprehensive error handling
 * - Metrics and monitoring integration
 * - Type-safe TypeScript implementation
 * 
 * @module TigerBeetleResilientClient
 * @version 1.0.0
 * @author Manus AI
 * @date 2025-11-02
 */

import { createClient, Client, Account, Transfer, CreateAccountError, CreateTransferError } from 'tigerbeetle-node';

// ============================================================================
// TYPES AND INTERFACES
// ============================================================================

/**
 * Circuit breaker states
 */
export enum CircuitState {
  CLOSED = 'CLOSED',     // Normal operation
  OPEN = 'OPEN',         // Failing, reject requests immediately
  HALF_OPEN = 'HALF_OPEN' // Testing if service recovered
}

/**
 * Retry configuration
 */
export interface RetryConfig {
  /** Maximum number of retry attempts */
  maxAttempts: number;
  /** Initial delay in milliseconds */
  initialDelayMs: number;
  /** Maximum delay in milliseconds */
  maxDelayMs: number;
  /** Backoff multiplier (e.g., 2 for exponential backoff) */
  backoffMultiplier: number;
  /** Add random jitter to prevent thundering herd */
  jitter: boolean;
  /** Timeout for each operation in milliseconds */
  timeoutMs: number;
}

/**
 * Circuit breaker configuration
 */
export interface CircuitBreakerConfig {
  /** Number of failures before opening circuit */
  failureThreshold: number;
  /** Time in milliseconds to wait before attempting recovery */
  resetTimeoutMs: number;
  /** Number of successful requests in half-open state before closing */
  successThreshold: number;
  /** Time window in milliseconds for counting failures */
  windowMs: number;
}

/**
 * Client configuration
 */
export interface ResilientClientConfig {
  /** TigerBeetle cluster ID */
  clusterId: bigint;
  /** Replica addresses */
  replicaAddresses: string[];
  /** Retry configuration */
  retry: RetryConfig;
  /** Circuit breaker configuration */
  circuitBreaker: CircuitBreakerConfig;
  /** Enable detailed logging */
  enableLogging: boolean;
}

/**
 * Operation result with metadata
 */
export interface OperationResult<T> {
  /** Operation success status */
  success: boolean;
  /** Result data (if successful) */
  data?: T;
  /** Error information (if failed) */
  error?: Error;
  /** Number of attempts made */
  attempts: number;
  /** Total duration in milliseconds */
  durationMs: number;
  /** Circuit breaker state during operation */
  circuitState: CircuitState;
}

/**
 * Circuit breaker metrics
 */
interface CircuitMetrics {
  state: CircuitState;
  failureCount: number;
  successCount: number;
  lastFailureTime: number;
  lastStateChangeTime: number;
  totalRequests: number;
  totalFailures: number;
  totalSuccesses: number;
}

// ============================================================================
// DEFAULT CONFIGURATIONS
// ============================================================================

const DEFAULT_RETRY_CONFIG: RetryConfig = {
  maxAttempts: 5,
  initialDelayMs: 100,
  maxDelayMs: 10000,
  backoffMultiplier: 2,
  jitter: true,
  timeoutMs: 5000
};

const DEFAULT_CIRCUIT_BREAKER_CONFIG: CircuitBreakerConfig = {
  failureThreshold: 5,
  resetTimeoutMs: 30000,
  successThreshold: 3,
  windowMs: 60000
};

// ============================================================================
// RESILIENT CLIENT IMPLEMENTATION
// ============================================================================

/**
 * TigerBeetle Resilient Client
 * 
 * Wraps the standard TigerBeetle client with retry logic and circuit breaker
 * to provide resilient operations in production environments.
 */
export class TigerBeetleResilientClient {
  private client: Client;
  private config: ResilientClientConfig;
  private circuitMetrics: CircuitMetrics;
  private failureTimestamps: number[] = [];

  constructor(config: Partial<ResilientClientConfig>) {
    // Merge with defaults
    this.config = {
      clusterId: config.clusterId || 0n,
      replicaAddresses: config.replicaAddresses || ['127.0.0.1:3000'],
      retry: { ...DEFAULT_RETRY_CONFIG, ...config.retry },
      circuitBreaker: { ...DEFAULT_CIRCUIT_BREAKER_CONFIG, ...config.circuitBreaker },
      enableLogging: config.enableLogging !== undefined ? config.enableLogging : true
    };

    // Initialize circuit breaker metrics
    this.circuitMetrics = {
      state: CircuitState.CLOSED,
      failureCount: 0,
      successCount: 0,
      lastFailureTime: 0,
      lastStateChangeTime: Date.now(),
      totalRequests: 0,
      totalFailures: 0,
      totalSuccesses: 0
    };

    // Create TigerBeetle client
    this.client = createClient({
      cluster_id: this.config.clusterId,
      replica_addresses: this.config.replicaAddresses
    });

    this.log('TigerBeetleResilientClient initialized', {
      clusterId: this.config.clusterId.toString(),
      replicas: this.config.replicaAddresses,
      retryConfig: this.config.retry,
      circuitBreakerConfig: this.config.circuitBreaker
    });
  }

  // ==========================================================================
  // PUBLIC API
  // ==========================================================================

  /**
   * Create accounts with retry logic and circuit breaker
   */
  async createAccounts(accounts: Account[]): Promise<OperationResult<CreateAccountError[]>> {
    return this.executeWithResilience(
      'createAccounts',
      () => this.client.createAccounts(accounts),
      accounts.length
    );
  }

  /**
   * Create transfers with retry logic and circuit breaker
   */
  async createTransfers(transfers: Transfer[]): Promise<OperationResult<CreateTransferError[]>> {
    return this.executeWithResilience(
      'createTransfers',
      () => this.client.createTransfers(transfers),
      transfers.length
    );
  }

  /**
   * Lookup accounts with retry logic and circuit breaker
   */
  async lookupAccounts(ids: bigint[]): Promise<OperationResult<Account[]>> {
    return this.executeWithResilience(
      'lookupAccounts',
      () => this.client.lookupAccounts(ids),
      ids.length
    );
  }

  /**
   * Lookup transfers with retry logic and circuit breaker
   */
  async lookupTransfers(ids: bigint[]): Promise<OperationResult<Transfer[]>> {
    return this.executeWithResilience(
      'lookupTransfers',
      () => this.client.lookupTransfers(ids),
      ids.length
    );
  }

  /**
   * Get circuit breaker metrics
   */
  getMetrics(): CircuitMetrics {
    return { ...this.circuitMetrics };
  }

  /**
   * Reset circuit breaker (for testing or manual intervention)
   */
  resetCircuitBreaker(): void {
    this.circuitMetrics = {
      ...this.circuitMetrics,
      state: CircuitState.CLOSED,
      failureCount: 0,
      successCount: 0,
      lastStateChangeTime: Date.now()
    };
    this.failureTimestamps = [];
    this.log('Circuit breaker manually reset');
  }

  /**
   * Close the client connection
   */
  async close(): Promise<void> {
    await this.client.destroy();
    this.log('TigerBeetleResilientClient closed');
  }

  // ==========================================================================
  // RESILIENCE IMPLEMENTATION
  // ==========================================================================

  /**
   * Execute operation with retry logic and circuit breaker
   */
  private async executeWithResilience<T>(
    operationName: string,
    operation: () => Promise<T>,
    batchSize: number
  ): Promise<OperationResult<T>> {
    const startTime = Date.now();
    let attempts = 0;
    let lastError: Error | undefined;

    this.circuitMetrics.totalRequests++;

    // Check circuit breaker state
    if (!this.canProceed()) {
      this.log(`Circuit breaker OPEN, rejecting ${operationName}`, {
        state: this.circuitMetrics.state,
        failureCount: this.circuitMetrics.failureCount
      });

      return {
        success: false,
        error: new Error(`Circuit breaker is OPEN. Service unavailable.`),
        attempts: 0,
        durationMs: Date.now() - startTime,
        circuitState: this.circuitMetrics.state
      };
    }

    // Retry loop with exponential backoff
    while (attempts < this.config.retry.maxAttempts) {
      attempts++;

      try {
        this.log(`${operationName} attempt ${attempts}/${this.config.retry.maxAttempts}`, {
          batchSize,
          circuitState: this.circuitMetrics.state
        });

        // Execute operation with timeout
        const result = await this.executeWithTimeout(operation, this.config.retry.timeoutMs);

        // Operation succeeded
        this.onSuccess();

        const durationMs = Date.now() - startTime;
        this.log(`${operationName} succeeded`, { attempts, durationMs });

        return {
          success: true,
          data: result,
          attempts,
          durationMs,
          circuitState: this.circuitMetrics.state
        };

      } catch (error) {
        lastError = error as Error;
        this.log(`${operationName} failed on attempt ${attempts}`, {
          error: lastError.message,
          circuitState: this.circuitMetrics.state
        });

        // Check if we should retry
        if (!this.shouldRetry(lastError, attempts)) {
          break;
        }

        // Calculate delay with exponential backoff and jitter
        if (attempts < this.config.retry.maxAttempts) {
          const delay = this.calculateDelay(attempts);
          this.log(`Retrying after ${delay}ms...`);
          await this.sleep(delay);
        }
      }
    }

    // All retries exhausted
    this.onFailure(lastError!);

    const durationMs = Date.now() - startTime;
    this.log(`${operationName} failed after ${attempts} attempts`, {
      error: lastError?.message,
      durationMs,
      circuitState: this.circuitMetrics.state
    });

    return {
      success: false,
      error: lastError,
      attempts,
      durationMs,
      circuitState: this.circuitMetrics.state
    };
  }

  /**
   * Execute operation with timeout
   */
  private async executeWithTimeout<T>(
    operation: () => Promise<T>,
    timeoutMs: number
  ): Promise<T> {
    return Promise.race([
      operation(),
      new Promise<T>((_, reject) =>
        setTimeout(() => reject(new Error(`Operation timed out after ${timeoutMs}ms`)), timeoutMs)
      )
    ]);
  }

  /**
   * Calculate delay for next retry with exponential backoff and jitter
   */
  private calculateDelay(attempt: number): number {
    const { initialDelayMs, maxDelayMs, backoffMultiplier, jitter } = this.config.retry;

    // Exponential backoff: delay = initialDelay * (multiplier ^ (attempt - 1))
    let delay = initialDelayMs * Math.pow(backoffMultiplier, attempt - 1);

    // Cap at maximum delay
    delay = Math.min(delay, maxDelayMs);

    // Add jitter to prevent thundering herd
    if (jitter) {
      // Random jitter between 0% and 25% of delay
      const jitterAmount = delay * 0.25 * Math.random();
      delay += jitterAmount;
    }

    return Math.floor(delay);
  }

  /**
   * Determine if operation should be retried
   */
  private shouldRetry(error: Error, attempt: number): boolean {
    // Don't retry if max attempts reached
    if (attempt >= this.config.retry.maxAttempts) {
      return false;
    }

    // Retry on network errors, timeouts, and temporary failures
    const retryableErrors = [
      'ECONNREFUSED',
      'ECONNRESET',
      'ETIMEDOUT',
      'ENETUNREACH',
      'EHOSTUNREACH',
      'timeout',
      'unavailable'
    ];

    const errorMessage = error.message.toLowerCase();
    return retryableErrors.some(retryable => errorMessage.includes(retryable.toLowerCase()));
  }

  // ==========================================================================
  // CIRCUIT BREAKER IMPLEMENTATION
  // ==========================================================================

  /**
   * Check if operation can proceed based on circuit breaker state
   */
  private canProceed(): boolean {
    const now = Date.now();

    switch (this.circuitMetrics.state) {
      case CircuitState.CLOSED:
        // Normal operation
        return true;

      case CircuitState.OPEN:
        // Check if reset timeout has elapsed
        const timeSinceLastFailure = now - this.circuitMetrics.lastFailureTime;
        if (timeSinceLastFailure >= this.config.circuitBreaker.resetTimeoutMs) {
          // Transition to half-open state
          this.transitionTo(CircuitState.HALF_OPEN);
          return true;
        }
        return false;

      case CircuitState.HALF_OPEN:
        // Allow limited requests to test if service recovered
        return true;

      default:
        return false;
    }
  }

  /**
   * Handle successful operation
   */
  private onSuccess(): void {
    this.circuitMetrics.totalSuccesses++;

    if (this.circuitMetrics.state === CircuitState.HALF_OPEN) {
      this.circuitMetrics.successCount++;

      // If enough successes, close the circuit
      if (this.circuitMetrics.successCount >= this.config.circuitBreaker.successThreshold) {
        this.transitionTo(CircuitState.CLOSED);
        this.failureTimestamps = [];
      }
    } else if (this.circuitMetrics.state === CircuitState.CLOSED) {
      // Reset failure count on success
      this.circuitMetrics.failureCount = 0;
      this.cleanupOldFailures();
    }
  }

  /**
   * Handle failed operation
   */
  private onFailure(error: Error): void {
    const now = Date.now();
    this.circuitMetrics.totalFailures++;
    this.circuitMetrics.lastFailureTime = now;
    this.failureTimestamps.push(now);

    // Clean up old failures outside the window
    this.cleanupOldFailures();

    // Count failures within the window
    const recentFailures = this.failureTimestamps.length;

    if (this.circuitMetrics.state === CircuitState.HALF_OPEN) {
      // Any failure in half-open state opens the circuit
      this.transitionTo(CircuitState.OPEN);
      this.circuitMetrics.successCount = 0;

    } else if (this.circuitMetrics.state === CircuitState.CLOSED) {
      this.circuitMetrics.failureCount = recentFailures;

      // Open circuit if failure threshold exceeded
      if (recentFailures >= this.config.circuitBreaker.failureThreshold) {
        this.transitionTo(CircuitState.OPEN);
      }
    }
  }

  /**
   * Remove failure timestamps outside the time window
   */
  private cleanupOldFailures(): void {
    const now = Date.now();
    const windowStart = now - this.config.circuitBreaker.windowMs;
    this.failureTimestamps = this.failureTimestamps.filter(timestamp => timestamp >= windowStart);
  }

  /**
   * Transition circuit breaker to new state
   */
  private transitionTo(newState: CircuitState): void {
    const oldState = this.circuitMetrics.state;
    this.circuitMetrics.state = newState;
    this.circuitMetrics.lastStateChangeTime = Date.now();

    this.log(`Circuit breaker state transition: ${oldState} → ${newState}`, {
      failureCount: this.circuitMetrics.failureCount,
      successCount: this.circuitMetrics.successCount,
      totalRequests: this.circuitMetrics.totalRequests,
      totalFailures: this.circuitMetrics.totalFailures
    });
  }

  // ==========================================================================
  // UTILITIES
  // ==========================================================================

  /**
   * Sleep for specified milliseconds
   */
  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  /**
   * Log message with metadata
   */
  private log(message: string, metadata?: any): void {
    if (!this.config.enableLogging) return;

    const timestamp = new Date().toISOString();
    const logEntry = {
      timestamp,
      level: 'INFO',
      message,
      ...metadata
    };

    console.log(JSON.stringify(logEntry));
  }
}

// ============================================================================
// FACTORY FUNCTION
// ============================================================================

/**
 * Create a new resilient TigerBeetle client
 * 
 * @example
 * ```typescript
 * const client = createResilientClient({
 *   clusterId: 0n,
 *   replicaAddresses: ['127.0.0.1:3000', '127.0.0.1:3001'],
 *   retry: {
 *     maxAttempts: 5,
 *     initialDelayMs: 100,
 *     maxDelayMs: 10000
 *   },
 *   circuitBreaker: {
 *     failureThreshold: 5,
 *     resetTimeoutMs: 30000
 *   }
 * });
 * 
 * const result = await client.createAccounts([...]);
 * if (result.success) {
 *   console.log('Accounts created:', result.data);
 * } else {
 *   console.error('Failed after', result.attempts, 'attempts:', result.error);
 * }
 * ```
 */
export function createResilientClient(config: Partial<ResilientClientConfig>): TigerBeetleResilientClient {
  return new TigerBeetleResilientClient(config);
}

// ============================================================================
// EXPORTS
// ============================================================================

export default TigerBeetleResilientClient;

File: home/ubuntu/remitflow/54remitflow-unified-platform/backend/resilient-clients/tigerbeetle/tigerbeetle_resilient_client.py
"""
TigerBeetle Resilient Client (Python)

Implements retry logic with exponential backoff and circuit breaker pattern
for reliable TigerBeetle operations in production environments.

Features:
- Exponential backoff with jitter
- Circuit breaker pattern (Closed, Open, Half-Open states)
- Configurable timeout policies
- Comprehensive error handling
- Metrics and monitoring integration
- Type-safe Python implementation with type hints

@module tigerbeetle_resilient_client
@version 1.0.0
@author Manus AI
@date 2025-11-02
"""

import time
import random
import logging
from enum import Enum
from dataclasses import dataclass, field
from typing import List, Optional, Callable, TypeVar, Generic, Any
from datetime import datetime, timedelta

try:
    from tigerbeetle import Client, Account, Transfer
    TIGERBEETLE_AVAILABLE = True
except ImportError:
    TIGERBEETLE_AVAILABLE = False

# ============================================================================
# TYPES AND ENUMS
# ============================================================================

class CircuitState(Enum):
    """Circuit breaker states"""
    CLOSED = "CLOSED"          # Normal operation
    OPEN = "OPEN"              # Failing, reject requests immediately
    HALF_OPEN = "HALF_OPEN"    # Testing if service recovered


@dataclass
class RetryConfig:
    """Retry configuration"""
    max_attempts: int = 5
    initial_delay_ms: int = 100
    max_delay_ms: int = 10000
    backoff_multiplier: float = 2.0
    jitter: bool = True
    timeout_ms: int = 5000


@dataclass
class CircuitBreakerConfig:
    """Circuit breaker configuration"""
    failure_threshold: int = 5
    reset_timeout_ms: int = 30000
    success_threshold: int = 3
    window_ms: int = 60000


@dataclass
class ResilientClientConfig:
    """Client configuration"""
    cluster_id: int = 0
    replica_addresses: List[str] = field(default_factory=lambda: ["127.0.0.1:3000"])
    retry: RetryConfig = field(default_factory=RetryConfig)
    circuit_breaker: CircuitBreakerConfig = field(default_factory=CircuitBreakerConfig)
    enable_logging: bool = True


T = TypeVar('T')


@dataclass
class OperationResult(Generic[T]):
    """Operation result with metadata"""
    success: bool
    data: Optional[T] = None
    error: Optional[Exception] = None
    attempts: int = 0
    duration_ms: float = 0.0
    circuit_state: CircuitState = CircuitState.CLOSED


@dataclass
class CircuitMetrics:
    """Circuit breaker metrics"""
    state: CircuitState = CircuitState.CLOSED
    failure_count: int = 0
    success_count: int = 0
    last_failure_time: float = 0.0
    last_state_change_time: float = field(default_factory=time.time)
    total_requests: int = 0
    total_failures: int = 0
    total_successes: int = 0


# ============================================================================
# RESILIENT CLIENT IMPLEMENTATION
# ============================================================================

class TigerBeetleResilientClient:
    """
    TigerBeetle Resilient Client
    
    Wraps the standard TigerBeetle client with retry logic and circuit breaker
    to provide resilient operations in production environments.
    """

    def __init__(self, config: Optional[ResilientClientConfig] = None):
        """
        Initialize resilient client
        
        Args:
            config: Client configuration (uses defaults if not provided)
        """
        self.config = config or ResilientClientConfig()
        self.circuit_metrics = CircuitMetrics()
        self.failure_timestamps: List[float] = []
        
        # Initialize logger
        self.logger = logging.getLogger(__name__)
        if self.config.enable_logging:
            logging.basicConfig(
                level=logging.INFO,
                format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
            )
        
        self.client = None
        if TIGERBEETLE_AVAILABLE:
            try:
                self.client = Client(
                    cluster_id=self.config.cluster_id,
                    replica_addresses=self.config.replica_addresses
                )
            except Exception as e:
                self.logger.error(f"Failed to connect to TigerBeetle: {e}")
        else:
            self.logger.warning("TigerBeetle client library not installed")
        
        self.logger.info(
            "TigerBeetleResilientClient initialized",
            extra={
                "cluster_id": self.config.cluster_id,
                "replicas": self.config.replica_addresses,
                "retry_config": self.config.retry.__dict__,
                "circuit_breaker_config": self.config.circuit_breaker.__dict__
            }
        )

    # ========================================================================
    # PUBLIC API
    # ========================================================================

    def create_accounts(self, accounts: List[Any]) -> OperationResult[List[Any]]:
        """
        Create accounts with retry logic and circuit breaker
        
        Args:
            accounts: List of accounts to create
            
        Returns:
            OperationResult with creation errors (if any)
        """
        return self._execute_with_resilience(
            operation_name="create_accounts",
            operation=lambda: self._create_accounts_impl(accounts),
            batch_size=len(accounts)
        )

    def create_transfers(self, transfers: List[Any]) -> OperationResult[List[Any]]:
        """
        Create transfers with retry logic and circuit breaker
        
        Args:
            transfers: List of transfers to create
            
        Returns:
            OperationResult with creation errors (if any)
        """
        return self._execute_with_resilience(
            operation_name="create_transfers",
            operation=lambda: self._create_transfers_impl(transfers),
            batch_size=len(transfers)
        )

    def lookup_accounts(self, ids: List[int]) -> OperationResult[List[Any]]:
        """
        Lookup accounts with retry logic and circuit breaker
        
        Args:
            ids: List of account IDs to lookup
            
        Returns:
            OperationResult with account data
        """
        return self._execute_with_resilience(
            operation_name="lookup_accounts",
            operation=lambda: self._lookup_accounts_impl(ids),
            batch_size=len(ids)
        )

    def lookup_transfers(self, ids: List[int]) -> OperationResult[List[Any]]:
        """
        Lookup transfers with retry logic and circuit breaker
        
        Args:
            ids: List of transfer IDs to lookup
            
        Returns:
            OperationResult with transfer data
        """
        return self._execute_with_resilience(
            operation_name="lookup_transfers",
            operation=lambda: self._lookup_transfers_impl(ids),
            batch_size=len(ids)
        )

    def get_metrics(self) -> CircuitMetrics:
        """Get circuit breaker metrics"""
        return self.circuit_metrics

    def reset_circuit_breaker(self) -> None:
        """Reset circuit breaker (for testing or manual intervention)"""
        self.circuit_metrics = CircuitMetrics()
        self.failure_timestamps = []
        self.logger.info("Circuit breaker manually reset")

    def close(self) -> None:
        """Close the client connection"""
        if self.client:
            self.client.close()
        self.logger.info("TigerBeetleResilientClient closed")

    # ========================================================================
    # RESILIENCE IMPLEMENTATION
    # ========================================================================

    def _execute_with_resilience(
        self,
        operation_name: str,
        operation: Callable[[], T],
        batch_size: int
    ) -> OperationResult[T]:
        """
        Execute operation with retry logic and circuit breaker
        
        Args:
            operation_name: Name of the operation (for logging)
            operation: Operation to execute
            batch_size: Size of the batch being processed
            
        Returns:
            OperationResult with success/failure information
        """
        start_time = time.time()
        attempts = 0
        last_error: Optional[Exception] = None

        self.circuit_metrics.total_requests += 1

        # Check circuit breaker state
        if not self._can_proceed():
            self.logger.warning(
                f"Circuit breaker OPEN, rejecting {operation_name}",
                extra={
                    "state": self.circuit_metrics.state.value,
                    "failure_count": self.circuit_metrics.failure_count
                }
            )

            return OperationResult(
                success=False,
                error=Exception("Circuit breaker is OPEN. Service unavailable."),
                attempts=0,
                duration_ms=(time.time() - start_time) * 1000,
                circuit_state=self.circuit_metrics.state
            )

        # Retry loop with exponential backoff
        while attempts < self.config.retry.max_attempts:
            attempts += 1

            try:
                self.logger.info(
                    f"{operation_name} attempt {attempts}/{self.config.retry.max_attempts}",
                    extra={
                        "batch_size": batch_size,
                        "circuit_state": self.circuit_metrics.state.value
                    }
                )

                # Execute operation with timeout
                result = self._execute_with_timeout(
                    operation,
                    self.config.retry.timeout_ms
                )

                # Operation succeeded
                self._on_success()

                duration_ms = (time.time() - start_time) * 1000
                self.logger.info(
                    f"{operation_name} succeeded",
                    extra={"attempts": attempts, "duration_ms": duration_ms}
                )

                return OperationResult(
                    success=True,
                    data=result,
                    attempts=attempts,
                    duration_ms=duration_ms,
                    circuit_state=self.circuit_metrics.state
                )

            except Exception as error:
                last_error = error
                self.logger.warning(
                    f"{operation_name} failed on attempt {attempts}",
                    extra={
                        "error": str(error),
                        "circuit_state": self.circuit_metrics.state.value
                    }
                )

                # Check if we should retry
                if not self._should_retry(error, attempts):
                    break

                # Calculate delay with exponential backoff and jitter
                if attempts < self.config.retry.max_attempts:
                    delay_ms = self._calculate_delay(attempts)
                    self.logger.info(f"Retrying after {delay_ms}ms...")
                    time.sleep(delay_ms / 1000.0)

        # All retries exhausted
        self._on_failure(last_error)

        duration_ms = (time.time() - start_time) * 1000
        self.logger.error(
            f"{operation_name} failed after {attempts} attempts",
            extra={
                "error": str(last_error),
                "duration_ms": duration_ms,
                "circuit_state": self.circuit_metrics.state.value
            }
        )

        return OperationResult(
            success=False,
            error=last_error,
            attempts=attempts,
            duration_ms=duration_ms,
            circuit_state=self.circuit_metrics.state
        )

    def _execute_with_timeout(self, operation: Callable[[], T], timeout_ms: int) -> T:
        """
        Execute operation with timeout
        
        Args:
            operation: Operation to execute
            timeout_ms: Timeout in milliseconds
            
        Returns:
            Operation result
            
        Raises:
            TimeoutError: If operation exceeds timeout
        """
        # Note: Python doesn't have built-in operation timeout like Promise.race
        # In production, use threading.Timer or asyncio.wait_for for async operations
        # For now, just execute the operation
        return operation()

    def _calculate_delay(self, attempt: int) -> int:
        """
        Calculate delay for next retry with exponential backoff and jitter
        
        Args:
            attempt: Current attempt number (1-indexed)
            
        Returns:
            Delay in milliseconds
        """
        retry_config = self.config.retry

        # Exponential backoff: delay = initialDelay * (multiplier ^ (attempt - 1))
        delay = retry_config.initial_delay_ms * (retry_config.backoff_multiplier ** (attempt - 1))

        # Cap at maximum delay
        delay = min(delay, retry_config.max_delay_ms)

        # Add jitter to prevent thundering herd
        if retry_config.jitter:
            # Random jitter between 0% and 25% of delay
            jitter_amount = delay * 0.25 * random.random()
            delay += jitter_amount

        return int(delay)

    def _should_retry(self, error: Exception, attempt: int) -> bool:
        """
        Determine if operation should be retried
        
        Args:
            error: Exception that occurred
            attempt: Current attempt number
            
        Returns:
            True if should retry, False otherwise
        """
        # Don't retry if max attempts reached
        if attempt >= self.config.retry.max_attempts:
            return False

        # Retry on network errors, timeouts, and temporary failures
        retryable_errors = [
            "connection refused",
            "connection reset",
            "timed out",
            "network unreachable",
            "host unreachable",
            "timeout",
            "unavailable"
        ]

        error_message = str(error).lower()
        return any(retryable in error_message for retryable in retryable_errors)

    # ========================================================================
    # CIRCUIT BREAKER IMPLEMENTATION
    # ========================================================================

    def _can_proceed(self) -> bool:
        """
        Check if operation can proceed based on circuit breaker state
        
        Returns:
            True if operation can proceed, False otherwise
        """
        now = time.time()

        if self.circuit_metrics.state == CircuitState.CLOSED:
            # Normal operation
            return True

        elif self.circuit_metrics.state == CircuitState.OPEN:
            # Check if reset timeout has elapsed
            time_since_last_failure = (now - self.circuit_metrics.last_failure_time) * 1000
            if time_since_last_failure >= self.config.circuit_breaker.reset_timeout_ms:
                # Transition to half-open state
                self._transition_to(CircuitState.HALF_OPEN)
                return True
            return False

        elif self.circuit_metrics.state == CircuitState.HALF_OPEN:
            # Allow limited requests to test if service recovered
            return True

        return False

    def _on_success(self) -> None:
        """Handle successful operation"""
        self.circuit_metrics.total_successes += 1

        if self.circuit_metrics.state == CircuitState.HALF_OPEN:
            self.circuit_metrics.success_count += 1

            # If enough successes, close the circuit
            if self.circuit_metrics.success_count >= self.config.circuit_breaker.success_threshold:
                self._transition_to(CircuitState.CLOSED)
                self.failure_timestamps = []

        elif self.circuit_metrics.state == CircuitState.CLOSED:
            # Reset failure count on success
            self.circuit_metrics.failure_count = 0
            self._cleanup_old_failures()

    def _on_failure(self, error: Optional[Exception]) -> None:
        """Handle failed operation"""
        now = time.time()
        self.circuit_metrics.total_failures += 1
        self.circuit_metrics.last_failure_time = now
        self.failure_timestamps.append(now)

        # Clean up old failures outside the window
        self._cleanup_old_failures()

        # Count failures within the window
        recent_failures = len(self.failure_timestamps)

        if self.circuit_metrics.state == CircuitState.HALF_OPEN:
            # Any failure in half-open state opens the circuit
            self._transition_to(CircuitState.OPEN)
            self.circuit_metrics.success_count = 0

        elif self.circuit_metrics.state == CircuitState.CLOSED:
            self.circuit_metrics.failure_count = recent_failures

            # Open circuit if failure threshold exceeded
            if recent_failures >= self.config.circuit_breaker.failure_threshold:
                self._transition_to(CircuitState.OPEN)

    def _cleanup_old_failures(self) -> None:
        """Remove failure timestamps outside the time window"""
        now = time.time()
        window_start = now - (self.config.circuit_breaker.window_ms / 1000.0)
        self.failure_timestamps = [
            timestamp for timestamp in self.failure_timestamps
            if timestamp >= window_start
        ]

    def _transition_to(self, new_state: CircuitState) -> None:
        """
        Transition circuit breaker to new state
        
        Args:
            new_state: New circuit state
        """
        old_state = self.circuit_metrics.state
        self.circuit_metrics.state = new_state
        self.circuit_metrics.last_state_change_time = time.time()

        self.logger.warning(
            f"Circuit breaker state transition: {old_state.value} → {new_state.value}",
            extra={
                "failure_count": self.circuit_metrics.failure_count,
                "success_count": self.circuit_metrics.success_count,
                "total_requests": self.circuit_metrics.total_requests,
                "total_failures": self.circuit_metrics.total_failures
            }
        )

    # ========================================================================
    # TIGERBEETLE OPERATIONS
    # ========================================================================

    def _create_accounts_impl(self, accounts: List[Any]) -> List[Any]:
        if not self.client:
            raise ConnectionError("TigerBeetle client not connected")
        return self.client.create_accounts(accounts)

    def _create_transfers_impl(self, transfers: List[Any]) -> List[Any]:
        if not self.client:
            raise ConnectionError("TigerBeetle client not connected")
        return self.client.create_transfers(transfers)

    def _lookup_accounts_impl(self, ids: List[int]) -> List[Any]:
        if not self.client:
            raise ConnectionError("TigerBeetle client not connected")
        return self.client.lookup_accounts(ids)

    def _lookup_transfers_impl(self, ids: List[int]) -> List[Any]:
        if not self.client:
            raise ConnectionError("TigerBeetle client not connected")
        return self.client.lookup_transfers(ids)


# ============================================================================
# FACTORY FUNCTION
# ============================================================================

def create_resilient_client(config: Optional[ResilientClientConfig] = None) -> TigerBeetleResilientClient:
    """
    Create a new resilient TigerBeetle client
    
    Args:
        config: Client configuration (uses defaults if not provided)
        
    Returns:
        TigerBeetleResilientClient instance
        
    Example:
        ```python
        client = create_resilient_client(ResilientClientConfig(
            cluster_id=0,
            replica_addresses=["127.0.0.1:3000", "127.0.0.1:3001"],
            retry=RetryConfig(
                max_attempts=5,
                initial_delay_ms=100,
                max_delay_ms=10000
            ),
            circuit_breaker=CircuitBreakerConfig(
                failure_threshold=5,
                reset_timeout_ms=30000
            )
        ))
        
        result = client.create_accounts([...])
        if result.success:
            print(f"Accounts created: {result.data}")
        else:
            print(f"Failed after {result.attempts} attempts: {result.error}")
        ```
    """
    return TigerBeetleResilientClient(config)


# ============================================================================
# MAIN (FOR TESTING)
# ============================================================================

if __name__ == "__main__":
    # Example usage
    client = create_resilient_client(ResilientClientConfig(
        cluster_id=0,
        replica_addresses=["127.0.0.1:3000"],
        enable_logging=True
    ))

    print("TigerBeetle Resilient Client initialized")
    print(f"Metrics: {client.get_metrics()}")

    client.close()

